design/encryption-spi.md

```diff
@@ -0,0 +1,166 @@
+>
+> (4) Administrator might want to decrypt all data.
+> This could be used to migrate data back to plaintext or to move to another encryption provider.
+> It could be used to recover from unseen issues.
+>
+> Supporting the above use cases would require much wider changes than presented in this document.
+> For example, there could be a "framework" to support marking which data fields are sensitive, making it possible to query and iterate over all sensitive fields in the database.
+
+### Interface
+
+```java
+public interface EncryptionServiceProvider {
+    Optional<byte[]> encrypt(byte[] plaintext);
+    Optional<byte[]> decrypt(byte[] ciphertext);
+}
+```
+
+### Supported fields
+
+Following fields are proposed to be encrypted by the SPI.
+Fields can be included incrementally in the future, and not all fields need to be covered at once.
+
+* Client secret generated by Keycloak.
+* Client secret configured by administrator for identity brokering.
+* External tokens stored in the database by Keycloak when "Store tokens" is configured for identity brokering.
+* Realm keys generated by Keycloak.
+* OTP shared secrets generated by Keycloak.
+* LDAP credentials configured by administrator.
+* SMTP credentials configured by administrator.
+
+There is some overlap with the Vault SPI.
+It is not anticipated that both the Vault SPI and Encryption SPI will be used simultaneously.
+The justification for the overlap are discussed in the Motivation section.
+
+Encryption SPI cannot be used for fields that need to remain searchable in the database.
+This can include fields like username, email, or other fields that can be considered as privacy-sensitive but need to be searchable.
+
+
+### Performance
+
+> **TODO**: Q: What is the expected performance impact?
+>
+> Based on the current proposal, the encryption service is invoked every time a secret is stored or retrieved from the database.
+>
+> There is no performance issue while the secret is in use, since it's kept in plaintext in memory.
+> However, if secrets are frequently retrieved or stored and the encryption service is remote, the performance impact could be significant.
+>
+> Considering the current usage patterns and in-memory caching, will the performance be acceptable?
+> For example, could this approach cause a spike in requests to the encryption service when the server starts up?
+> Should we consider a more advanced solution where most encryption is handled locally, reducing the number of calls to the encryption service?
+
+### Default implementation
+
+The encryption SPI is disabled by default.
+When disabled there is no impact on the existing functionality of Keycloak or the format of the data stored in the database.
+
+> **TODO**: Q: Should Keycloak include default provider for Encryption SPI?
+>
+> Some possible alternatives:
```

Comment:
Having an SPI without an implementation is not very elegant; it would more or less make it useless to most folks. Additionally it would be impossible to test fully, as well as do any sort of benchmarking, which really would be required for something like this.

HashiCorp is a defacto choice, but less so in the open source now that it has changed its license. So, not really sure what a default implementation (or implementations) would be.

Reply:
Based on your other comment we could in theory have two implementations; one that stores in DB directly as encrypted values (that should of course be able to receive its own encryption keys from a secret, or from HashiCorp, etc.); and another that stores in an external vault like HashiCorp.